pyro/poutine/enumerate_messenger.py

```diff
@@ -5,13 +5,63 @@
 from .runtime import _ENUM_ALLOCATOR
 
 
+def _tmc_sample(msg):
+    dist, num_samples = msg["fn"], msg["infer"].get("num_samples")
+
+    # find batch dims that aren't plate dims
+    batch_shape = [1] * len(dist.batch_shape)
+    for f in msg["cond_indep_stack"]:
+        if f.vectorized:
+            batch_shape[f.dim] = f.size
+    batch_shape = tuple(batch_shape)
+
+    # sample a batch
+    sample_shape = (num_samples,) if batch_shape == dist.batch_shape else ()
+    fat_sample = dist(sample_shape=sample_shape)  # TODO thin before sampling
```

Comment:
Sampling like this is inefficient - we should really collapse the batch dimensions in the distribution parameters before sampling rather than taking polynomially too many samples and dropping most of them, but that will require quite a bit of distribution plumbing.